services/venice-server/src/main/java/com/linkedin/venice/grpc/GrpcIoRequestProcessor.java

```diff
@@ -0,0 +1,212 @@
+package com.linkedin.venice.grpc;
+
+import static com.linkedin.venice.listener.ReadQuotaEnforcementHandler.INVALID_REQUEST_RESOURCE_MSG;
+import static com.linkedin.venice.listener.ReadQuotaEnforcementHandler.QuotaEnforcementResult.ALLOWED;
+import static com.linkedin.venice.listener.ReadQuotaEnforcementHandler.SERVER_OVER_CAPACITY_MSG;
+import static io.netty.handler.codec.http.HttpResponseStatus.NOT_FOUND;
+import static io.netty.handler.codec.http.HttpResponseStatus.OK;
+
+import com.google.protobuf.ByteString;
+import com.linkedin.davinci.listener.response.ReadResponse;
+import com.linkedin.venice.HttpConstants;
+import com.linkedin.venice.exceptions.VeniceException;
+import com.linkedin.venice.listener.ReadQuotaEnforcementHandler;
+import com.linkedin.venice.listener.ReadQuotaEnforcementHandler.QuotaEnforcementResult;
+import com.linkedin.venice.listener.ServerStatsContext;
+import com.linkedin.venice.listener.StorageReadRequestHandler;
+import com.linkedin.venice.listener.request.RouterRequest;
+import com.linkedin.venice.protocols.MultiGetResponse;
+import com.linkedin.venice.protocols.SingleGetResponse;
+import com.linkedin.venice.protocols.VeniceServerResponse;
+import com.linkedin.venice.response.VeniceReadResponseStatus;
+import com.linkedin.venice.stats.ServerHttpRequestStats;
+import com.linkedin.venice.utils.LatencyUtils;
+import io.grpc.stub.StreamObserver;
+import io.netty.handler.codec.http.HttpResponseStatus;
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+
+
+public class GrpcIoRequestProcessor {
+  private static final Logger LOGGER = LogManager.getLogger(GrpcIoRequestProcessor.class);
+  private final ReadQuotaEnforcementHandler readQuotaEnforcementHandler;
+  private final StorageReadRequestHandler storageReadRequestHandler;
+
+  public GrpcIoRequestProcessor(GrpcServiceDependencies services) {
+    this.readQuotaEnforcementHandler = services.getReadQuotaEnforcementHandler();
+    this.storageReadRequestHandler = services.getStorageReadRequestHandler();
+  }
+
+  public void processRequest(GrpcRequestContext requestContext) {
+    RouterRequest request = requestContext.getRouterRequest();
+    QuotaEnforcementResult result =
+        readQuotaEnforcementHandler != null ? readQuotaEnforcementHandler.enforceQuota(request) : null;
+    // If the request is allowed, hand it off to the storage read request handler
+    if (result == null || result == ALLOWED) {
+      GrpcStorageResponseHandlerCallback callback = GrpcStorageResponseHandlerCallback.create(requestContext);
+      storageReadRequestHandler.queueIoRequestForAsyncProcessing(request, callback);
+      return;
+    }
```

Comment:
Instead of these null checks, can we make the `ReadQuotaEnforcementHandler` never null, meaning it's either going to contain a real handler, or else a "no op quota enforcement handler" which always returns `ALLOWED`? This behavior is decided at construction time and will never change afterwards, so there is no point in having the conditional checks on the hot path. Being never null also makes the code more readable (IMHO).

Reply:
Yeah good point. Added `NoOpReadQuotaEnforcementHandler`